powerauth-push-server/src/main/java/io/getlime/push/service/PushMessageSenderService.java

```diff
@@ -332,19 +372,55 @@ private void sendMessageToAndroid(final FcmClient fcmClient, final PushMessageBo
             notification.setTag(pushMessageBody.getCategory());
             request.setNotification(notification);
         }
-        final ListenableFuture<ResponseEntity<String>> future = fcmClient.exchange(request);
+        final ListenableFuture<ResponseEntity<FcmSendResponse>> future = fcmClient.exchange(request);
 
-        future.addCallback(new ListenableFutureCallback<ResponseEntity<String>>() {
+        future.addCallback(new ListenableFutureCallback<ResponseEntity<FcmSendResponse>>() {
             @Override
             public void onFailure(Throwable throwable) {
                 Logger.getLogger(PushMessageSenderService.class.getName()).log(Level.SEVERE, "Notification rejected by the FCM gateway:" + throwable.getLocalizedMessage());
                 Logger.getLogger(PushMessageSenderService.class.getName()).log(Level.SEVERE, throwable.getLocalizedMessage());
-                callback.didFinishSendingMessage(PushSendingCallback.Result.FAILED);
+                callback.didFinishSendingMessage(PushSendingCallback.Result.FAILED, null);
             }
+
             @Override
-            public void onSuccess(ResponseEntity<String> stringResponseEntity) { // TODO: Implement processing of the response body
-                Logger.getLogger(PushMessageSenderService.class.getName()).log(Level.INFO, stringResponseEntity.getBody());
-                callback.didFinishSendingMessage(PushSendingCallback.Result.OK);
+            public void onSuccess(ResponseEntity<FcmSendResponse> stringResponseEntity) {
+                for (FcmResult fcmResult : stringResponseEntity.getBody().getFcmResults()) {
+                    //no issues, straight sending
+                    if (stringResponseEntity.getBody().getFailure() + stringResponseEntity.getBody().getCanonicalIds() == 0) {
```

Comment:
This code is IMO not correct, Google says in their documentation:

> If the value of failure and canonical_ids is 0, it's not necessary to parse the remainder of the response. Otherwise, we recommend that you iterate through the results field and do the following for each object in that list...

It means you should first do this "if" and if the values are non-zero, you should start processing the sent items...